src/math/bvh.js

```diff
@@ -0,0 +1,456 @@
+                        if (this.minDist == null) {
+                            this.minDist = dist;
+                        }
+                        this.minDist = Math.min(this.minDist, dist);
+                    }
+                }
+                if (stackPtr === 0) {
+                    break;
+                } else {
+                    node = stack[--stackPtr];
+                    continue;
+                }
+            }
+            let child1 = this.bvhNode[node.leftFirst];
+            let child2 = this.bvhNode[node.leftFirst + 1];
+            let dist1 = this.intersectAABB(ray, child1.aabbMin, child1.aabbMax);
+            let dist2 = this.intersectAABB(ray, child2.aabbMin, child2.aabbMax);
+            if (dist1 > dist2) {
+                [dist1, dist2] = [dist2, dist1];
+                [child1, child2] = [child2, child1];
+            }
+            if (dist1 === Infinity) {
+                if (stackPtr === 0) {
+                    break;
+                } else {
+                    node = stack[--stackPtr];
+                }
+            } else {
+                node = child1;
+                if (dist2 !== Infinity) {
+                    stack[stackPtr++] = child2;
+                }
+            }
+        }
+    }
+
+    /**
+     * Calculates whether a ray intersects an aabb and returns the distance between the ray origin and aabb if it does
+     * Ray.rdx, rdy and rdz must be calculated prior to this
+     *
+     * @param {Ray} ray - The ray to be traced
+     * @param {Vec3} bmin - The min corner of the aabb
+     * @param {Vec3} bmax - The max corner of the aabb
+     * @returns {number} The distance between the ray origin and the aabb, Infinity if there is no intersection
+     * @ignore
+     */
+    intersectAABB(ray, bmin, bmax) {
+        const tx1 = (bmin.x - ray.origin.x) * ray.rDx;
+        const tx2 = (bmax.x - ray.origin.x) * ray.rDx;
+        let tmin = Math.min(tx1, tx2);
+        let tmax = Math.max(tx1, tx2);
+        const ty1 = (bmin.y - ray.origin.y) * ray.rDy;
+        const ty2 = (bmax.y - ray.origin.y) * ray.rDy;
+        tmin = Math.max(tmin, Math.min(ty1, ty2));
+        tmax = Math.min(tmax, Math.max(ty1, ty2));
+        const tz1 = (bmin.z - ray.origin.z) * ray.rDz;
+        const tz2 = (bmax.z - ray.origin.z) * ray.rDz;
+        tmin = Math.max(tmin, Math.min(tz1, tz2));
+        tmax = Math.min(tmax, Math.max(tz1, tz2));
+        if (tmax >= tmin  && tmax > 0 && (this.minDist == null || (this.minDist != null && tmin < this.minDist))) {
+            return tmin;
+        }
+        return Infinity;
+    }
+}
+
+
+/**
+ * Returns a vector with the minimum x, y and z coordinates of the two vectors passed in
+ *
+ * @param {Vec3} a - First Vector
+ * @param {Vec3} b - Second Vector
+ * @param {Vec3} [c] - The vector to output the result to - default is new Vec3()
+ * @returns {Vec3} Returns the vector result
+ * @ignore
+ */
+function vec3Min(a, b, c = new Vec3()) {
```

Comment:
I'd add these to Vec3 class, called min2 (we already have min). See Vec3.add vs Vec3.add2

Reply:
Added new min2 and max2 functions, and changed all lines in bvh to use those as appropriate, as well as a new temporary internal: `tempVec1`